skimage/morphology/_extrema_cy.pyx

```diff
@@ -0,0 +1,191 @@
+#cython: cdivision=True
+#cython: boundscheck=False
+#cython: nonecheck=False
+#cython: wraparound=False
+
+cimport numpy as cnp
+from libc.stdlib cimport malloc, realloc, free
+
+
+ctypedef fused dtype_t:
+    cnp.uint8_t
+    cnp.uint16_t
+    cnp.double_t
+
+
+ctypedef Py_ssize_t QueueItem
+
+
+cdef:
+    struct Queue:
+        # A queue whose items can be restored after consumption.
+        QueueItem* _buffer_ptr
+        Py_ssize_t _buffer_size, _index_valid, _index_consumed
+
+    void q_init(Queue* self, Py_ssize_t buffer_size) nogil:
+        """Initialize the queue and its buffer."""
+        self._buffer_ptr = <QueueItem*>malloc(buffer_size * sizeof(QueueItem))
+        if not self._buffer_ptr:
+            with gil:
+                raise MemoryError("couldn't allocate buffer")
+        self._buffer_size = buffer_size
+        self._index_consumed = -1
+        self._index_valid = -1
+
+    void q_free_buffer(Queue* self) nogil:
+        """Free the buffer of the queue."""
+        free(self._buffer_ptr)
+
+    inline void q_restore(Queue* self) nogil:
+        """Restore all consumed items to the queue."""
+        self._index_consumed = -1
+
+    inline void q_clear(Queue* self) nogil:#
+        """Remove all items in the queue."""
+        self._index_consumed = -1
+        self._index_valid = -1
+
+    inline void q_push(Queue* self, QueueItem* item_ptr) nogil:
+        """Enqueue a new item."""
+        self._index_valid += 1
+        if self._buffer_size <= self._index_valid:
+            _q_grow_buffer(self)
+        self._buffer_ptr[self._index_valid] = item_ptr[0]
+
+    inline unsigned char q_pop(Queue* self, QueueItem* item_ptr) nogil:
+        """Dequeue / consume an item."""
+        if 0 <= self._index_consumed + 1 <= self._index_valid:
+            self._index_consumed += 1
+            item_ptr[0] = self._buffer_ptr[self._index_consumed]
+            return 1
+        return 0
+
+    void _q_grow_buffer(Queue* self) nogil:
+        """Double the memory used for the buffer."""
+        cdef QueueItem* new_buffer
+
+        # TODO prevent integer overflow!
+        self._buffer_size *= 2
+        new_buffer_ptr = <QueueItem*>realloc(
```

Comment:
I expect that there is a constant or variable that stores the maximal possible size of `Py_ssize_t` but I can't find it.